File: 2020/2.php
<?php

require('libs/core.php');

function validate_1($passwords) {
    $total = 0;
    foreach ($passwords as $i => $line) {
        if (preg_match('/^(\d+)\-(\d+) (\w): (\w+)$/', $line, $matches)) {
            $count = substr_count($matches[4], $matches[3]);
            $valid = ($count >= $matches[1]) && ($count <= $matches[2]);
            if ($valid) {
                $total++;
            }
        }
    }
    print "\n{$total} are valid\n";
}

function validate_2($passwords) {
    $total = 0;
    foreach ($passwords as $i => $line) {
        if (preg_match('/^(\d+)\-(\d+) (\w): (\w+)$/', $line, $matches)) {
            $char1 = $matches[4][$matches[1] - 1];
            $char2 = $matches[4][$matches[2] - 1];
            $valid1 = $char1 == $matches[3];
            $valid2 = $char2 == $matches[3];
            $valid = ($valid1 xor $valid2);
            if ($valid) {
                $total++;
            }
        }
    }
    print "\n{$total} are valid\n";
}

$passwords = load_data("2.txt");

validate_1($passwords);

validate_2($passwords);

File: 2020/python/2.py
import core
import re

def validate_1(numbers):
	total = 0
	for line in numbers:
		matches = re.search('^(\d+)\-(\d+) (\w): (\w+)$', line) # None if not found
		if matches == None:
			continue
		else:
			count = matches.group(4).count(matches.group(3))
			valid = (count >= int(matches.group(1))) and (count <= int(matches.group(2)))
			if valid:
				total += 1
	print("%s are valid" % total)

def validate_2(numbers):
	total = 0
	for line in numbers:
		matches = re.search('^(\d+)\-(\d+) (\w): (\w+)$', line)
		if matches == None:
			continue
		else:
			char1 = matches.group(4)[int(matches.group(1)) - 1]
			char2 = matches.group(4)[int(matches.group(2)) - 1]
			valid1 = char1 == matches.group(3)
			valid2 = char2 == matches.group(3)
			valid = valid1 ^ valid2
			if valid:
				total += 1
	print("%s are valid" % total)

passwords = core.load_data("2.txt")

validate_1(passwords);

validate_2(passwords);
